invisible_cities/reco/paolina_functions.py

```diff
@@ -100,8 +100,19 @@ def find_extrema(distance : Dict[Voxel, Dict[Voxel, float]]) -> Tuple[Voxel, Vox
     return (first, last)
 
 
-def length(track_graph):
-    return len(shortest_paths(track_graph))
+def length(distance : Dict[Voxel, Dict[Voxel, float]]) -> float:
+    """Calculate the length of a track."""
+    if not distance:
+        raise NoVoxels
+    if len(distance) == 1:
+        return 0.
+    max_distance = 0.
```

Comment:
This is essentially copy-pasted from `find_extrema`.

We do not do copy-n-paste code reuse in IC: it is an evil practice which leads to confusion and to wasting lots of time and effort.

If you need to reuse some code, wrap it into a convenient abstraction, such as a function, and use that in multiple places.

In this case you should extract the loop from `find_extrema` into a function, and call that function in both `find_extrema` and `length`.

Reply:
Ok.

Reply:
I am also a bit uncomfortable with the argument of length and find_extrema. From an external user's point of view, I would expect to look for the length or the extremes of a track, not of a dictionary of distances. I understand that calling all_pairs_dijkstra_path_length() every time I ask for a length is maybe unnecessarily time-consuming? Is this the reason why the find_extrema function has the overmentioned dictionary as an argument?

Reply:
The primary reason is probably decomposition and testability. Optimization might well be a secondary reason.

In `paolina_functions.py` at least, `find_extrema` is *always* called just after `shortest_paths`.

I think it would be reasonable to add a function which calculates the extrema directly from a track: it would simply call `shortest_paths` and `find_extrema` in succession.

You might want to rename `find_extrema` to something more clunky and use the original name for the new one which accepts tracks. And then it would be reasonable to do something similar for `length`.